src/core/border_router/routing_manager.cpp

```diff
@@ -373,22 +471,50 @@ bool RoutingManager::ContainsPrefix(const Ip6::Prefix &aPrefix, const Ip6::Prefi
     return ret;
 }
 
-const Ip6::Prefix *RoutingManager::EvaluateOnLinkPrefix(void) const
+const Ip6::Prefix *RoutingManager::EvaluateOnLinkPrefix(void)
 {
-    const Ip6::Prefix *newOnLinkPrefix = nullptr;
+    const Ip6::Prefix *newOnLinkPrefix      = nullptr;
+    const Ip6::Prefix *smallestOnLinkPrefix = nullptr;
+    uint8_t            onLinkPrefixNum      = 0;
 
-    // We don't evaluate on-link prefix if we are doing
-    // Router Discovery or we have already discovered some
-    // on-link prefixes.
+    // We don't evaluate on-link prefix if we are doing Router Solicitation.
     VerifyOrExit(!mRouterSolicitTimer.IsRunning());
-    if (IsValidOnLinkPrefix(mDiscoveredOnLinkPrefix))
+
+    for (uint8_t i = 0; i < mDiscoveredPrefixNum; ++i)
     {
-        otLogInfoBr("EvaluateOnLinkPrefix: there is already on-link prefix %s on interface %u",
-                    mDiscoveredOnLinkPrefix.ToString().AsCString(), mInfraIfIndex);
-        ExitNow();
+        if (!mDiscoveredPrefixes[i].mIsOnLinkPrefix)
+        {
+            continue;
+        }
+
+        ++onLinkPrefixNum;
+
+        if (smallestOnLinkPrefix == nullptr ||
+            IsPrefixSmallerThan(mDiscoveredPrefixes[i].mPrefix, *smallestOnLinkPrefix))
+        {
+            smallestOnLinkPrefix = &mDiscoveredPrefixes[i].mPrefix;
+        }
+    }
+
+    // We start advertising our local on-link prefix if there is no existing one.
+    if (onLinkPrefixNum == 0)
+    {
+        if (AddExternalRoute(mLocalOnLinkPrefix, OT_ROUTE_PREFERENCE_MED) == OT_ERROR_NONE)
+        {
+            newOnLinkPrefix = &mLocalOnLinkPrefix;
+        }
     }
+    // When an application-specific on-link prefix is received and it is bigger than the
+    // advertised prefix, we will not remove the advertised prefix. In this case, there
+    // will be two on-link prefixes on the infra link. But all BRs will still converge to
+    // the same smallest on-link prefix and the application-specific prefix is not used.
+    else if (mAdvertisedOnLinkPrefix != nullptr && *smallestOnLinkPrefix != *mAdvertisedOnLinkPrefix)
+    {
+        otLogInfoBr("EvaluateOnLinkPrefix: there is already smaller on-link prefix %s on interface %u",
+                    smallestOnLinkPrefix->ToString().AsCString(), mInfraIfIndex);
 
-    newOnLinkPrefix = &mLocalOnLinkPrefix;
+        RemoveExternalRoute(*mAdvertisedOnLinkPrefix);
+    }
 
 exit:
     return newOnLinkPrefix;
```

Comment:
The `newOnLinkPrefix` would be null except when there is no existing on-link prefix in the list.

I see that this is then used to update the `mAdvertisedOnLinkPrefix` from `EvaluateRoutingPolicy()`. 

If we see that there are other prefixes but ours is smaller, should we not continue to to keep  `mAdvertisedOnLinkPrefix` as before (and send it in our RA?)

Reply:
Good Catch!